### DISCUSSION TITLE
Why C++ / native code?

### DISCUSSION THREAD

**Discussion by mikevercoelen** [jpudysz/react-native-unistyles#132] [General]
Hi, this looks like a very promising project to solve some outstanding problems with RN StyleSheet.

And apologies in advance if this comes off as a negative, I'm generally just curious.

But why was C++ / native code chosen? I'm only seeing downsides and I'm not seeing things you can't do at regular runtime.

Was this for performance reasons or for features that aren't possible at runtime?

**Comment by jpudysz** [comment]
Hello, no worries, no offense taken 😇.

I’ve chosen to use JSI and C++ for a few reasons. The main advantage is that it allows me to gradually transfer more responsibilities to the native side, squeezing out extra milliseconds from each stylesheet.

Unistyles offers a unique architecture. With each version, I plan to reveal more APIs to expose unexplored areas of mobile styling.

One such feature is the content size category: https://reactnativeunistyles.vercel.app/reference/content-size-category/, along with some properties on UnistylesRuntime. There’s more on the way! I believe that in the future you will be able to replace some libraries by only using Unistyles. 

Leveraging native features on native platforms is crucial. With unistyles 2.0, I’ve already managed to reduce the number of code, offering easy support for dark mode with just a single line 🙏

Every event in Unistyles is processed through C++. Based on various factors, I decide when to re-render components. I don’t rely on any React Native events for this.

In the future, I plan to use more host objects and some shadow nodes with a custom renderer to fully harness the power of React Native. It’s a long journey, but I’m committed to supporting it. I’m also implementing unistyles in my commercial projects, and some are already using it in production.

There are other benefits, but in short, I’m using it to introduce a unique architecture with a custom-made event system and to expand the possibilities of integrating more deeply with the native world.

For me, using native APIs is not a downside, as we’re operating in a native environment. I hope this answers your question ☺️.

**Reply by mikevercoelen** [reply]
It sounds like you've taken the native route not because you had to, but because you assume it's faster / you're enjoying writing native code (which I fully understand) have you benchmarked and proven that C++ is faster, as it comes with the bridge overhead?

I'm still not sure, are there any features that are not possible through plain old React Native lib? i.e. 

- Media queries: https://reactnative.dev/docs/usewindowdimensions
- Theme: https://reactnative.dev/docs/usecolorscheme

etc.

As for downsides, here's a few that come to my mind right away:

- It makes it harder for others to contribute + increased complexity
- It needs a full build / breaks compatibility with plain old Expo Go (which is not important, but it's a downside because it's way faster for people to try out the library without having to create a custom dev client)
- If `unistyles` is updated, it can't be distributed through OTA's, it needs a full app review

All in all, I'm still not seeing any problems that can't be solved, in a performant way using what is already available in RN.

**Reply by jpudysz** [reply]
> It sounds like you've taken the native route not because you had to, but because you assume it's faster / you're enjoying writing native code (which I fully understand)

Not only because of that, also because i have endless possiblities with access to native APIs.

> have you benchmarked and proven that C++ is faster, as it comes with the bridge overhead?

Yup, it's not always faster eg. for initialization it's slower a little bit (~1ms), but for larger amount of stylesheets it's faster.
You can check my findings here: https://reactnativeunistyles.vercel.app/start/benchmarks/

> I'm still not sure, are there any features that are not possible through plain old React Native lib?

I am using more events. I used these hooks in version 1. An example is the content size category mentioned earlier. Also, it's worth mentioning that I aim to bring more accessibility features to Unistyles.

> As for downsides, here's a few that come to my mind right away:
 - for contribution I agree, it's harder for someone who has no knowledge in native code, but I'm trying to structure everything as friendly as possible. Also, I already had some good contributions from the community
 - regarding Expo Go, it's just a simple playground and it shouldn't be used in production. I see no issue with bulding your code with one command
 - regarding OTA - yeah, it's also a downside, but it's the same story with reanimated, gesture handler and many many more libraries. I'm trying to release it in predictable manner

I believe everyone needs to ask themselves a question: Is Unistyles worth using in their project? They need to weigh the benefits and downsides. I'm not pushing it onto everyone's codebase, and I'm aware that there are other alternatives with many fans. For me, it's great to use it, as I have control over every style and can build a UI Kit for each project with almost the full speed of a StyleSheet.

**Reply by mikevercoelen** [reply]
Got it, thanks!
